csharp/src/Apache.Arrow/Arrays/BinaryArray.cs

```diff
@@ -66,87 +66,158 @@ protected BuilderBase(IArrowType dataType)
+            /// <param name="value">Enumerable collection of bytes to add.</param>
+            /// <returns>Returns the builder (for fluent-style composition).</returns>
+            public TBuilder Append(IEnumerable<byte> value)
             {
-                foreach (byte[] arr in values)
+                if (value == null)
                 {
-                    if (arr == null)
-                    {
-                        AppendNull();
-                        continue;
-                    }
-                    int len = ValueBuffer.Length;
-                    ValueOffsets.Append(Offset);
-                    ValueBuffer.Append(arr);
-                    ValidityBuffer.Append(true);
-                    Offset += ValueBuffer.Length - len;
+                    return AppendNull();
                 }
 
+                // Note: by looking at the length of the value buffer before and after, we avoid having to iterate
+                // through the enumerable multiple times to get both length and contents.
+                int priorLength = ValueBuffer.Length;
+                ValueBuffer.AppendRange(value);
+                int valueLength = ValueBuffer.Length - priorLength;
+                Offset += valueLength;
+                ValidityBuffer.Append(true);
+                ValueOffsets.Append(Offset);
                 return Instance;
             }
 
+            /// <summary>
+            /// Append an enumerable collection of single-byte values to the array.
+            /// </summary>
+            /// <remarks>
+            /// Note that this method appends multiple values, each of which is a single byte.  If a single value is
+            /// to be added, use the <see cref="Append(IEnumerable{byte})"/> method instead.
+            /// </remarks>
+            /// <param name="values">Single-byte values to add.</param>
+            /// <returns>Returns the builder (for fluent-style composition).</returns>
             public TBuilder AppendRange(IEnumerable<byte> values)
             {
-                if (values == null)
+                foreach (byte b in values)
                 {
-                    return AppendNull();
+                    Append(b);
                 }
-                int len = ValueBuffer.Length;
-                ValueBuffer.AppendRange(values);
-                int valOffset = ValueBuffer.Length - len;
-                ValueOffsets.Append(Offset);
-                Offset += valOffset;
-                ValidityBuffer.Append(true);
+
+                return Instance;
+            }
+
+            /// <summary>
+            /// Append an enumerable collection of values to the array.
+            /// </summary>
+            /// <param name="values">Values to add.</param>
+            /// <returns>Returns the builder (for fluent-style composition).</returns>
+            public TBuilder AppendRange(IEnumerable<byte[]> values)
+            {
+                foreach (byte[] arr in values)
```

Comment:
This will throw a `NullReferenceException` if `values` is `null`.

Reply:
Fixed.